[WIP] additionnal lib/cleanup for French language to improve quality of inputs

CapitainFlam opened the pull request:
<H1>better french input for better french output ?</h1>

Hi ! 
⚠️ Firsts commits in GitHub, first pull request ever in a repo, so please neither shoot or shout at me 😅  ⚠️ 

This commit comes from different discussion, that I propose you to have a look to understand the context. 

- https://github.com/common-voice/commonvoice-fr/issues/21 _(...This issue should be partially solved with this pull)_
- https://github.com/common-voice/CorporaCreator/pull/87 _(...This is inspirational commit that I took some ideas from)_

and it is somehow connected to : 

- https://github.com/common-voice/commonvoice-fr/issues/20
- https://commonvoice.mozilla.org/fr/criteria
- https://discourse.mozilla.org/t/rfc-criteres-dacceptation-pour-la-validation-denregistrements/73063/12 
- https://discourse.mozilla.org/t/proportion-de-noms-propres-a-consonnance-etrangere/72887/26
- https://github.com/common-voice/common-voice/pull/3786
- https://github.com/common-voice/common-voice/issues/3792
- https://github.com/common-voice/cv-sentence-extractor/issues/92

**TL;DR: 
I'm trying to improve input quality for french sentence collector, to avoid 
**_A/_** garbage IN - garbage OUT, and 
_B/_ avoid to remove in the end (in the [CorporaCreator](https://github.com/common-voice/CorporaCreator)) the garbage that went through collecting, recording, review... to finally being dropped and not being included in the final release for training batchs.** Better to remove it as soon as sentence collector.

<H1>What have you done ? 😱 </h1>

To do so, I duplicated a EN to FR file in sentence collector, and modified it according to a previous job made by Nicolas Panel for CorporaCreator (and sadly not commited). 

According to discussion (links [here](https://github.com/common-voice/commonvoice-fr/issues/21#issuecomment-1242611833)... Did you follow it from the list above ?!), it's recommanded that I create a WIP pull request, to allow everyone to comment, throw tomatos and/or additionnal commits to it. 

footnote : it can be hard (...it IS hard !!!) to understand REGEX (REGular EXpressions). Do not hesitate to catch up with https://regex101.com/ to understand and test it.

CapitainFlam committed:
adding FR rules in languages and index.js

CapitainFlam committed:
adding many FR rules, transcoding acronyms, and removing numbers

...some of my first commits, and my first steps in JS.
reviews are welcome.
(...please neither shoot or shout at me ;-D )

Co-Authored-By: Nicolas Panel <2500584+nicolaspanel@users.noreply.github.com>

CapitainFlam committed:
Update fr.js

CapitainFlam committed:
Update fr.js

roman + century numerals,
roman numerals,
full convertion of all ACRONYMES

CapitainFlam committed:
Update fr.js

eleventh, ...
first, second,...
1/2 => 1 sur 2

drzraf commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+      .replace(/(^|\s)TVA(\s|\.|,|\?|!|$)/g, ' Taxe sur la Valeur Ajoutée ')
+      .replace(/(^|\s)UDI(\s|\.|,|\?|!|$)/g, ' Union des Démocrates Indépendants ')
+      .replace(/(^|\s)UMP(\s|\.|,|\?|!|$)/g, ' Union pour un Mouvement Populaire ')
+      .replace(/(^|\s)USA(\s|\.|,|\?|!|$)/g, ' Etats Unis d\'Amérique ')
```
Side note: Acronyms are a strange case : They are actually spelled as distincts letters (except for NUPES).

But we **don't** want "STT" to process `R.S.A.` as `erre et ça` **but we currently have nothing to train it that way**.

Should we expect a specific future dictionary or just not strip some of them in the first place (but remove the dots)?

MichaelKohler commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+	  .replace(/(^|\s)(XIX|XVIIII)(\s|\.|,|\?|!|$)/g, ' dix-neuf ')
+	  .replace(/(^|\s)XX(\s|\.|,|\?|!|$)/g, ' vingt ')
+ 	  .replace(/(^|\s)XXI(\s|\.|,|\?|!|$)/g, ' vingt-et-un ')
+ 	  .replace(/(^|\s)XXII(\s|\.|,|\?|!|$)/g, ' vingt-deux ')
```
I think we can remove all of the roman numerals, as they are not allowed from the beginning (see the abbreviation pattern in the validation file).

MichaelKohler commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+	  .replace(/(^|\s)7i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' septième ')
+	  .replace(/(^|\s)8i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' huitième ')
+	  .replace(/(^|\s)9i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' neuvième ')
+	  .replace(/(^|\s)10i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' dixième ')
```
These can be removed, as numbers are not allowed anyway in the Sentence Collector.

MichaelKohler commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+      .replace(/(^|\s)TVA(\s|\.|,|\?|!|$)/g, ' Taxe sur la Valeur Ajoutée ')
+      .replace(/(^|\s)UDI(\s|\.|,|\?|!|$)/g, ' Union des Démocrates Indépendants ')
+      .replace(/(^|\s)UMP(\s|\.|,|\?|!|$)/g, ' Union pour un Mouvement Populaire ')
+      .replace(/(^|\s)USA(\s|\.|,|\?|!|$)/g, ' Etats Unis d\'Amérique ')
```
These can be removed, as abbreviations are not allowed in the validation file.

CapitainFlam replied:
Well, I think I understand wrongly how it works. Let's solve this https://github.com/common-voice/sentence-collector/issues/636 issue first 😸

MichaelKohler commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+      .replace(/(^|\s)USA(\s|\.|,|\?|!|$)/g, ' Etats Unis d\'Amérique ')
+
+      //replace fraction 1/2 => '1 sur 2'
+     ¨.replace(/(^| )(\d+)(\s)?(\/)(\s)?(\d+)(\s|\.|,|\?|!|$)/g, '$2 sur $6')
```
Numbers are not allowed by the validation file.

CapitainFlam replied:
Shouldn't we keep it for "future-proof rules" and/or "bulk load rules if you don't have yours" ? 
Let's discuss it here : https://discourse.mozilla.org/t/sentence-collector-cleanup-before-export-vs-cleanup-on-upload/105411/15

MichaelKohler commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+	  //dates, digits and numbers fr-FR cleanup
+	  //todo : CONVERT TO TEXT instead of removing it
+	  .replace((^|\s)\d{1,2}\/\d{1,2}\/(\d{2}[^\d]|\d{4})(\s|$), ' ') //date format dd/mm/yy ou dd/mm/yyyy
+	  .replace((^|\s)\d{1,2}\/(\d{2}[^\d]|\d{4})(\s|$), ' ') //date format mm/yy ou mm/yyyy
```
Numbers are not allowed by the validation file.

MichaelKohler commented on server/lib/validation/languages/fr.js:
```diff
@@ -0,0 +1,32 @@
+    const words = tokenizeWords(sentence);
+    return words.length < MIN_WORDS || words.length > MAX_WORDS;
+  },
+  error: `Number of words must be between ${MIN_WORDS} and ${MAX_WORDS} (inclusive) - Le nombre de mot doit être entre ${MIN_WORDS} et ${MAX_WORDS} (inclus)`,
```
I probably would leave only the French error message in here (goes for all of them). I don't particularly see a benefit of having both languages shown in the UI.

MichaelKohler reviewed (changes requested):
> ⚠️ Firsts commits in GitHub, first pull request ever in a repo, so please neither shoot or shout at me 😅 ⚠️

No worries, you're doing fine :) Thanks for your contributions!

I have a few comments, but generally I think it's a good thing to update these files.

MichaelKohler commented:
Also note https://discourse.mozilla.org/t/sentence-collector-cleanup-before-export-vs-cleanup-on-upload/105411, though that probably doesn't matter too much for this here :)

MichaelKohler commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+	  //based on common-voice/CorporaCreator#87
+	  .replace(/(^|\s|\w)\/an(\s|\.|,|\?|!|$)/g, '$1 par an ')
+	  .replace(/(^|\s)km(\s|\.|,|\?|!|$)/g, ' kilomètres ')
+	  .replace(/%, ' pourcent ')
```
```suggestion
	  .replace(/%/, ' pourcent ')
```

CapitainFlam committed:
Update server/lib/cleanup/languages/fr.js

of course :)

Co-authored-by: Michael Kohler <me@michaelkohler.info>

CapitainFlam commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+      .replace(/(^|\s)USA(\s|\.|,|\?|!|$)/g, ' Etats Unis d\'Amérique ')
+
+      //replace fraction 1/2 => '1 sur 2'
+     ¨.replace(/(^| )(\d+)(\s)?(\/)(\s)?(\d+)(\s|\.|,|\?|!|$)/g, '$2 sur $6')
```
oh shoot ! you're right ! 
To solve it, we shall test numbers, and see "if it's in date [1..31]/[1..12] range, i shall be a date stamp, otherwise it's a fraction ?!"
Sometimes, only context will show us that "mettre 1/4 de litre de lait" will go to "mettre un avril de litre de lait"... But it shall be more readable than "mettre de litre de lait".

CapitainFlam replied:
Let's first wait for issue https://github.com/common-voice/sentence-collector/issues/636 resolution, 'coz I don't want to work for nothing :-/

CapitainFlam commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -10,7 +10,9 @@ function sortSentences(sentences) {
-      // no space after opening '(' or '['
+	  //caracters and space cleanup
+	  
+	  // no space after opening '(' or '['
```
Yep. Wrong configuration of Notepad++ on my side (Tab instead of SpaceSpaceSpaceSpace) 😅 
Will correct it next time.

CapitainFlam commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+	  //based on common-voice/CorporaCreator#87
+
+	  //Jean-Paul II
+	  .replace(/Jean-Paul II|Jean Paul II/g, 'Jean-Paul deux')
```
It was just the work of Nicolas. ...I think we can remove it. (It was based on the dump file of the CorporaCreator errors, it could explain this 'only 1 entry' list)

CapitainFlam committed:
Update fr.js

Notepad++ > Edit > transform TABS in SPACE.
Yeahhhhhh

CapitainFlam committed:
Update fr.js

as requested, only keeping FR and removing EN error description...
But adding EN comment, for ^FR readers sake.

CapitainFlam commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+	  .replace(/(^|\s)(XIX|XVIIII)(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' dix_neuvième siècle ')
+	  .replace(/(^|\s)XX(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' vingtième siècle ')
+	  .replace(/(^|\s)XXI(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' vingt-et-unième siècle ')
+	  .replace(/(^|\s)XXII(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' vingt-deuxième siècle ')
```
well, 
### First, 
I have my regex to work with : 
-- kudos to [gregory-r](https://stackoverflow.com/users/2872366/gregory-r) in [this post in stackoverflow](https://stackoverflow.com/a/49946096/20003021), licence [CC BY-SA](https://stackoverflow.com/help/licensing)

```javascript
function deromanize(roman) {
  var r = 0;
  // regular expressions to check if valid Roman Number.
  if (!/^M*(?:D?C{0,3}|C[MD])(?:L?X{0,3}|X[CL])(?:V?I{0,3}|I[XV])$/.test(roman))
    throw new Error('Invalid Roman Numeral.');

  roman.replace(/[MDLV]|C[MD]?|X[CL]?|I[XV]?/g, function(i) {
    r += {M:1000, CM:900, D:500, CD:400, C:100, XC:90, L:50, XL:40, X:10, IX:9, V:5, IV:4, I:1}[i]; 
  });

  return r;
}
```
So, I'll put it here to not forget it, because : 

### Second, 
I don't have a clue of how to implement a function in JS ! (I know Pascal and VB ✌️ , but never learned JS 😭... I have to jump in the rabbit hole 🕳️ 🐇  (it is [this way](https://www.youtube.com/watch?v=6IDT3MpSCKI)) )

...To however follow, either the white rabbit, or [crumbs](https://dictionary.cambridge.org/fr/dictionnaire/anglais/crumb), including future me, well, this link
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_function_as_the_replacement 
could be a good start to look at.

/me : jumps in the 🕳️

CapitainFlam replied:
Indeed, I don't understand that, collecting sentences, we can NOT "correct and clean" before "validation".

no problem, I'll wait for the issue  https://github.com/common-voice/sentence-collector/issues/636 😸

CapitainFlam commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+	  .replace(/(^|\s)7i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' septième ')
+	  .replace(/(^|\s)8i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' huitième ')
+	  .replace(/(^|\s)9i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' neuvième ')
+	  .replace(/(^|\s)10i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' dixième ')
```
I actually don't know how to import external library, I think it's a bit out of my league right now.

CapitainFlam committed:
Update fr.js

We do not remove PUNCTUATION (?!, etc.) after finding them, we just put them back  as we found them.

CapitainFlam commented on server/lib/cleanup/languages/fr.js:
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
+      .replace(/(^|\s)TVA(\s|\.|,|\?|!|$)/g, ' Taxe sur la Valeur Ajoutée ')
+      .replace(/(^|\s)UDI(\s|\.|,|\?|!|$)/g, ' Union des Démocrates Indépendants ')
+      .replace(/(^|\s)UMP(\s|\.|,|\?|!|$)/g, ' Union pour un Mouvement Populaire ')
+      .replace(/(^|\s)USA(\s|\.|,|\?|!|$)/g, ' Etats Unis d\'Amérique ')
```
Est ce que le commit https://github.com/common-voice/sentence-collector/pull/635/commits/aef428430ae01712fb491e09dae157b9b4b62c3b répond au problème ?

CapitainFlam committed:
Update fr.js

remplacing – (long hyphen) by - (short hyphen)

CapitainFlam commented on server/lib/validation/languages/fr.js:
```diff
@@ -0,0 +1,32 @@
+  error: 'Sentence should not contain numbers - les phrases ne doivent pas contenir de nombres',
+}, {
+  regex: /[<>+*#@%^[\]()/]/,
+  error: 'Sentence should not contain symbols - les phrases de doivent pas contenir de symboles \(\*, \#, \(, etc\)',
```
commit https://github.com/common-voice/sentence-collector/pull/635/commits/18fdf37f6554dc015b19c82cac1b5627a82e766d

CapitainFlam committed:
Update fr.js

Normalize ´ (french apostroph) into ' (usual apostroph)

CapitainFlam commented on server/lib/validation/languages/fr.js:
```diff
@@ -0,0 +1,32 @@
+  // as users wouldn't know how to pronounce the uppercase letters.
+  regex: /[A-Z]{2,}|[A-Z]+\.*[A-Z]+/,
+  error: 'Sentence should not contain abbreviations - Les phrases ne doivent pas contenir des abréviations ou sigles',
+}];
```
commit https://github.com/common-voice/sentence-collector/pull/635/commits/1a0cff5772ac2015ed8e2e90660069913fd75fac

CapitainFlam replied:
BTW, sources files like https://github.com/common-voice/commonvoice-fr/blob/master/CommonVoice-Data/data/debats-assemblee-nationale/20130718093000000.txt are using french apostrophe ?! 
(it only shows in "ui-monospace" code font, not the usual "Segoe UI" font. 
```
line 2 : J’appelle maintenant, dans le texte de la commission, les articles du projet de loi.
if I change it myself : 
line 2 : J'appelle maintenant, dans le texte de la commission, les articles du projet de loi.
```

CapitainFlam replied:
so... commit https://github.com/common-voice/sentence-collector/pull/635/commits/8d7d97340fb5e98c47228a8fd709d46d1968b7f1

CapitainFlam committed:
Update fr.js

adding an other apostroph found in original source document as https://github.com/common-voice/commonvoice-fr/blob/master/CommonVoice-Data/data/debats-assemblee-nationale/20130718093000000.txt

drzraf commented:
Any hope to get this in, in one shape or another?

MichaelKohler commented:
The Sentence Collector has now been integrated into the Common Voice platform. Therefore I'm archiving this project here. The validation files now live here and I'm sure it would still benefit from the validation rules being added there: https://github.com/common-voice/common-voice/tree/main/server/src/core/sentences/validation. Unfortunately moving this PR over there is way harder than manually recreating it. Would you mind creating a new PR for this? Sorry for the troubles.
